core/runtime/common/storage_code_provider.cpp

```diff
@@ -34,11 +35,14 @@ namespace kagome::runtime {
     if (last_state_root_ != state) {
       last_state_root_ = state;
 
-      auto hash = runtime_upgrade_tracker_->getLastCodeUpdateHash(state);
-      if (hash.has_value()) {
-        if (known_code_substitutes_->count(hash.value())) {
-          OUTCOME_TRY(code,
-                      chain_spec_->fetchCodeSubstituteByHash(hash.value()));
+      auto block_info =
+          runtime_upgrade_tracker_->getLastCodeUpdateBlockInfo(state);
+      if (block_info.has_value()) {
+        if (known_code_substitutes_->count(block_info.value().number)
```

Comment:
All such expressions seem pretty dangerous, it'd be easy to forget that you need to check both. Maybe normalize them to either number or hash, when possible, and then work only with entries of one of the types?

Reply:
IMO there is no elegant solution to that thing (actually code substitutes themselves are not elegant). We could also consider adding some helper function to code_substitutes.hpp that takes code substitutes and block info as input and returns true if provided block info matches any code substitute. But we may keep that way as well